File: ai-problems/657judgecircle.cpp
class Solution {
public:
    bool judgeCircle(string moves) {
        // O(n) time O1 space simulation approach
        int x=0, y=0;
        for(char move : moves) {
            if(move == 'L') x--;
            if(move == 'R') x++;
            if(move == 'D') y--;
            if(move == 'U') y++;
        }
        
        return (x == 0 && y ==0);
    }
};

File: ai-problems/657judgeCircle.py
"""
There is a robot starting at position (0, 0), the origin, on a 2D plane. Given a sequence of its moves, judge if this robot ends up at (0, 0) after it completes its moves.

The move sequence is represented by a string, and the character moves[i] represents its ith move. Valid moves are R (right), L (left), U (up), and D (down). If the robot returns to the origin after it finishes all of its moves, return true. Otherwise, return false.

Note: The way that the robot is "facing" is irrelevant. "R" will always make the robot move to the right once, "L" will always make it move left, etc. Also, assume that the magnitude of the robot's movement is the same for each move.

 

Example 1:

Input: moves = "UD"
Output: true
Explanation: The robot moves up once, and then down once. All moves have the same magnitude, so it ended up at the origin where it started. Therefore, we return true.
"""

class Solution:
    def judgeCircle(self, moves: str) -> bool:
        """2D sim approach"""
        x = y = 0
        
        for move in moves:
            if move == 'L': x-=1
            if move == 'R': x+=1
            if move == 'D': y-=1
            if move == 'U': y+=1
        
        return x == y == 0